src/jit/codegenxarch.cpp

```diff
@@ -3680,6 +3680,65 @@ void CodeGen::genStructPutArgRepMovs(GenTreePutArgStk* putArgNode, unsigned base
     genConsumePutStructArgStk(putArgNode, REG_RDI, REG_RSI, REG_RCX, baseVarNum);
     instGen(INS_r_movsb);
 }
+
+//------------------------------------------------------------------------
+// If any Vector3 args are on stack and they are not pass-by-ref, the upper 32bits
+// must be cleared to zeroes. The native compiler doesn't clear the upper bits
+// and there is no way to know if the caller is native or not. So, the upper
+// 32 bits of Vector argument on stack are always cleared to zero.
+#ifdef FEATURE_SIMD
+void    CodeGen::genClearStackVec3ArgUpperBits()
+{
+#ifdef DEBUG
+    if (verbose)
+        printf("*************** In genClearStackVec3ArgUpperBits()\n");
+#endif
+
+    assert(compiler->compGeneratingProlog);
+
+    unsigned varNum = 0;
+
+    for (LclVarDsc * varDsc = compiler->lvaTable;
+    varNum < compiler->lvaCount;
+        varNum++, varDsc++)
+    {
+        if (!varDsc->lvIsParam)
+        {
+            continue;
+        }
+
+        // Does var has simd12 type?
+        if (!varDsc->lvIsSIMDType() || varDsc->lvType != TYP_SIMD12)
+        {
+            continue;
+        }
+
+        if (!varDsc->lvIsRegArg)
+        {
+            // Clear the upper 32 bits by mov dword ptr [V_ARG_BASE+0xC], 0
+            getEmitter()->emitIns_S_I(
+                ins_Store(TYP_INT),
+                EA_4BYTE,
+                varNum,
+                genTypeSize(TYP_FLOAT) * 3,
+                0);
+
+        }
+        else {
+            regNumber argReg = varDsc->GetOtherArgReg();
+
+            // Clear the upper 32 bits by two shift instructions.
+            if (argReg != REG_STK)
```

Comment:
I believe that is true for x64 linux, but Siva's question makes the point that we should note in the comments that this logic depends on the argument being passed either fully in registers or fully on the stack.